### PR TITLE
Fixed URL schema for temporary panel

### LINKED ISSUE #377: File attributes functionality does not work on items in temporary panel
The FA action should change attributes for files on _base_ filesystem.

- Add FSTmp::StatSetAttr override passing the action to base FS.
- Rethink URL scheme for FSTmp to have consistency between every FSTmp object and its URL. This is maintained by Parze*() and FSTmp::Uri() functions.

### LINKED ISSUE #443: File view/edit history doesn't work for temp panel items
It would be useful to remember view/edit positions/history for files put on temp panel.

### PR DESCRIPTION
**Pull request by helgp**
- Fixed #443  - view history on temp panel
- Fixed #377 - attributes on temp panel

### CODE REVIEW

File: src/vfs/vfs-tmp.cpp [hunk 1]
```diff
@@ -362,11 +364,61 @@ int FSTmp::SetFileTime(FSPath& path, FSTime cTime, FSTime aTime, FSTime mTime, i
...
+// 2. Delegate file Uri to baseBS. Folder Uri is tmp:///folder/name#baseFSRootUri
+//   - appears to work 
+FSString FSTmp::Uri(FSPath& path)
+{
+	FSTmpNode* n = rootDir.findByFsPath(&path);
+	if (n && n->nodeType == FSTmpNode::NODE_FILE)
+		return baseFS->Uri(n->baseFSPath);
+	else
+	{
+		std::string ret(std::string("tmp://") + path.GetUtf8('/') + "#" + baseFS->Uri(rootPathName).GetUtf8());
```
Comment (corporateshark) [inline src/vfs/vfs-tmp.cpp]: Why ``path.GetUtf8('/')``? Why not just ``"/"``?

File: src/vfs/vfs-tmp.cpp [hunk 2]
```diff
@@ -362,11 +364,61 @@ int FSTmp::SetFileTime(FSPath& path, FSTime cTime, FSTime aTime, FSTime mTime, i
...
+	}
+}
+
+// Handles only folder URIs.
+// File URIs are not tmp://-based, but are baseFS-based
+// The URI is: tmp:///folder/name#baseFSRoot
+clPtr<FS> FSTmp::ParzeURI(const unicode_t* uri, FSPath& path, const std::vector<clPtr<FS>>& checkFS)
+{
+	std::string	uriUtf8(unicode_to_utf8_string(uri));
+	if (uriUtf8.compare(0, 6, "tmp://") == 0)
```
Comment (corporateshark) [inline src/vfs/vfs-tmp.cpp]: Please, use LUrlParser for this.

**Comment by viktor-podzigun** [comment]
For me #443 not working, at least on Windows. Please, note my last comment at #443 for desired behavior.

BTW, why not using base FS for URI always when on Temp panel? This way I think most of the cases will be solved. Temp panel is just a thin wrapper (references) around local FS.

**Comment by helgp** [comment]
>BTW, why not using base FS for URI always when on Temp panel? 

What would be the URL for a folder in TMPfs in your case?

**Comment by viktor-podzigun** [comment]
I would say the same as for local (system).

**Comment by helgp** [comment]
> I would say the same as for local (system).
Then, the "F5" operation to copy a file to a temporary panel would copy the file to the "local(system) FS" instead of copying it to the TMPfs.

**Comment by helgp** [comment]
There is still a bug in this:

View history on FSTmp works fine when entry for a viewed file already exists in registry, though it does not work when the file is first viewed on temp panel. 

Please, do not merge until I fix that.

**Comment by corporateshark** [comment]
Ok

**Comment by helgp** [comment]
> There is still a bug in this

An update on the bug #443.

It should be easier to fix the problem by changing the approach how view history records are stored in registry. 

Currently, the format for the file on local FS is:
DISK=2;PATH="\\0\\makefile";NAME="C:\\0\\makefile";VIEW_POS=0;TYPE=SYSTEM;IS_EDIT=0;
with special encoding rules for SMB/SFTP/FTP filesystems. See path-list.cpp/PathListFSToData() for the full set of rules. Extending this to TMPFs requres to put an extra case into the switch for the set of rules regarding TMPFs.

The suggestion is to store the values "DISK=", "PATH=", "NAME=", "TYPE=" in a single entry, named "
"URI", which is obtained from FS.Uri(FilePath) virtual function for current FS. URI is guaranteed to be unique for each file.

The URI approach is better, because it avoids a long switch through all file system types within PathListFSToData(). Adding new file system types (like archiveFS) will not need extra changes to the PathListFSToData() - as long as archiveFS provides consistent URI mapping with its Uri() method.

Also, the existing history implementation does not distinct between remote files with the same path on different servers. E.g. the files sftp://server1/etc/passwd, and sftp://server2/etc/passwd use same view history entry. This will be fixed with the proposed change to URI.

@viktor-podzigun, your word. Please. The history component is your domain. What do you think about the suggestion? If you agree, will you do it, or I will change the encoding schema for the history?

@corporateshark, the PR can be merged now. The #443 is not completely fixed with the change, though, based on the above, its fixing does not require touching FSTmp code.

**Comment by corporateshark** [comment]
@help do we have any possible regression with this? Was actually planning to release 0.19.1 in a couple of weeks.

**Comment by helgp** [comment]
I am unaware of any regression with the commit. I do believe that closing the #443 in the proposed way fits within the 2 weeks cycle without regression either.

**Comment by viktor-podzigun** [comment]
> Also, the existing history implementation does not distinct between remote files with the same path on different servers. E.g. the files sftp://server1/etc/passwd, and sftp://server2/etc/passwd use same view history entry.

I'm not agree, since entries are stored/fetched by NAME which is already unique and match to the URI (see path-list.cpp:324 - data.name = utf8_to_unicode(fs[0]->Uri(*path).GetUtf8())).

> Currently, the format for the file on local FS is:
DISK=2;PATH="\0\makefile";NAME="C:\0\makefile";VIEW_POS=0;TYPE=SYSTEM;IS_EDIT=0;
with special encoding rules for SMB/SFTP/FTP filesystems. See path-list.cpp/PathListFSToData() for the full set of rules. Extending this to TMPFs requres to put an extra case into the switch for the set of rules regarding TMPFs.

I don't like switch/case approach either, and I agree with you that we can improve this. But in case of TMPFs, which is now the only one which is not working properly because of mentioned missing case under the switch.

> @viktor-podzigun, your word. Please. The history component is your domain. What do you think about the suggestion? If you agree, will you do it, or I will change the encoding schema for the history?

I already proposed a solution by treating files on TMPFs as local (system) entries, for which we already have case under the switch :D So this way it will work without touching path-list.cpp and the use case described in #443 will also be supported.

I prefer this approach because its effortless.

**Comment by viktor-podzigun** [comment]
For quick solution that doesn't breaks anything but fixes #443 according to desired behavior, please, see #474

**Pull request merged**